src/lib.rs

```diff
@@ -27,10 +34,100 @@ fn optimise(
     Ok((z, delta, iterations))
 }
 
+fn inverse_odds_for_z(z: f64, prob_seq: &[f64]) -> Vec<f64> {
+    let y: Vec<f64> = prob_seq
+        .iter()
+        .map(|&p| f64::sqrt((z * p) + ((1.0 - z) * p * p)))
+        .collect();
+    let sum_y: f64 = y.iter().sum();
+    y.into_iter().map(|yy| yy * sum_y).collect()
+}
+
+struct ZSolve<'a> {
+    probabilities: &'a [f64],
+    overround: f64,
+}
+
+impl<'a> CostFunction for ZSolve<'a> {
+    type Param = f64;
+    type Output = f64;
+
+    fn cost(&self, z: &Self::Param) -> Result<Self::Output, Error> {
+        let sum_inverse_odds = inverse_odds_for_z(*z, &self.probabilities)
+            .iter()
+            .sum::<f64>();
+        Ok((sum_inverse_odds - (1.0 + self.overround)).abs())
+    }
+}
+
+fn solve_for_z(prob_seq: &[f64], overround: f64) -> f64 {
+    if overround < 0.0 {
+        panic!("Overround must be non-negative");
+    }
+
+    if overround > 1.0 {
+        panic!("Overround must be less than 1.0");
+    }
+
+    if overround == 0.0 {
+        return 0.0;
+    }
+
+    let cost = ZSolve {
+        probabilities: prob_seq,
+        overround,
+    };
+    let solver = BrentOpt::new(0., 1.).set_tolerance(f64::EPSILON.sqrt(), 1e-7);
+
+    let res = Executor::new(cost, solver)
+        .configure(|state| state.max_iters(100).target_cost(0.0))
+        .add_observer(SlogLogger::term(), ObserverMode::Always)
+        .run()
+        .unwrap();
+
+    res.state.best_param.unwrap()
+}
+
+#[pyfunction]
+fn calculate_implied_odds(probabilities: Vec<f64>, overround: f64) -> Vec<f64> {
```

Comment:
Should expose things like tolerance, and max iters.

Also, find a way to get stats out of the solver, and be able to toggle the logging on/off via parameter.

Reply:
maybe a flag to enable normalizing the input probabilities to 100%??